core/src/apps/solana/ui/format.py

```diff
@@ -0,0 +1,42 @@
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from ..transaction.instructions import Instruction
+
+
+def format_pubkey(_: Instruction, value: bytes | None) -> str:
+    from trezor.crypto import base58
+
+    if value is None:
+        raise ValueError  # should not be called with optional pubkey
+
+    return base58.encode(value)
+
+
+def format_lamports(_: Instruction, value: int) -> str:
+    from trezor.strings import format_amount
+
+    formatted = format_amount(value, decimals=9)
+    return f"{formatted} SOL"
+
+
+def format_token_amount(instruction: Instruction, value: int) -> str:
+    from trezor.strings import format_amount
+
+    decimals = instruction.decimals if instruction.decimals is not None else 0
```

Comment:
does this always work, i.e., is the `instruction` guaranteed to have `decimals` ?

Reply:
No, it's not really guaranteed although it is true now - any instruction using `format_token_amount` does have a decimals attr although it's not represented in the code.

I'm not sure what the best way to fix this is. One option is to return `None` from from `Instruction.__getattr` if the attr is not found. Then the current condition wouldn't cause problems anymore although that's IMO a bug prone solution.

Another is to simply wrap this line in a `try/except` and set `decimals = 0` in the `except` block. This could also be extracted into `Instruction.get_attr_with_default` for example (but extracting it would also require the attribute to be referenced by a string).

Another is to do `if "decimals" in instruction and instruction.decimals is not None` instead of the current condition - although I don't like the attribute being referenced by string.

I don't know of a simple (or perhaps any) way to represent this also in the type system somehow.

Reply:
like `assert hasattr(instruction, "decimals")  # enforced by build_solana_templates.py`

Reply:
> in which case would instruction.decimals exist but be None?

The `is None` check was there because I thought that it would work as a check for the instruction not having the decimals - the code is wrong (also in my comment), we don't expect decimals to exist and be None.

> and, do we ever expect to have an instruction that uses format_token_amount but does not have decimals ?

No.

> i'm leaning towards keeping this code as-is, but adding a static check to the template generator, to the tune of "if using format_token_amount, then check that there's also a decimals field"? (and referencing it in a comment here)

Could work. I'll take a look.

Reply:
Templates updated in https://github.com/trezor/trezor-firmware/pull/3359/commits/8a17e693a1ef0d1fc865458857c600adc3b4a2a3 and code updated in https://github.com/trezor/trezor-firmware/pull/3359/commits/4d9f08502322c4dcddfa096e7651fef99241ffc7.